intake_esm/core.py

```diff
@@ -82,6 +82,66 @@ def __init__(self, esmcol_path, progressbar=True, log_level='CRITICAL', **kwargs
         self.cdf_kwargs = None
         self.preprocess = None
         self.aggregate = None
+        self.sep = sep
+
+    def __len__(self):
+        return len(self.df)
+
+    def __getitem__(self, key):
+        # The canonical unique key is the path of an asset
+        # We also accept other aliases here
+        path_column_name = self._col_data['assets']['column_name']
+
+        # First, try the canonical unique key. This is faster than searching
+        results = self.df.loc[self.df[path_column_name] == key]
+
+        # Next, try aliases when the canonical key returned nothing
+        if results.empty:
+            columns = self.df.columns.tolist()
+            columns.remove(path_column_name)
+            key_parts = key.split(self.sep)
+            query = {k: v for k, v in zip(columns, key_parts) if v != '*'}
+            results = self.search(**query)
+
+        if len(results) == 1:
+
+            if isinstance(results, pd.DataFrame):
+                asset = _get_asset_info(self._col_data, results)
+            else:
+                asset = _get_asset_info(self._col_data, results.df)
+
+            asset = asset.popitem()
+            asset_path = asset[0]
+            data_format = asset[1][1]
+
+            if data_format == 'zarr':
+                driver = 'zarr'
+
+            else:
+                driver = 'netcdf'
+
+            # TODO: figure out how to pass other args such as storage_options, chunks, etc
+            entry = LocalCatalogEntry(
```

Comment:
I've backtracked a bit on my original design target here. What I'm thinking would be most useful (as a start) is to provide access to an intake Entry for each key typically found in `to_dataset_dict()`. So these two datasets would be equivalent:

```python
dset_dict = cat.to_dataset_dict()
ds = dset_dict['CMIP.CCCma.CanESM5.historical.Oyr.gn']

# and
cat['CMIP.CCCma.CanESM5.historical.Oyr.gn'].to_dask()
```

Then `to_dataset_dict()` becomes just a convenience method:

```python
def to_dataset_dict(self, **kwargs):
    return {key: self[key](**kwargs).to_dask() for key in self.keys()}
```

What we need right now is to understand how to take a group of rows from the catalog's dataframe, and construct a single `GroupedEntry` where the loading logic is implemented in intake-esm.  @andersy005, can you spec out what information this `GroupedEntry` would need to construct a single xarray Dataset?